bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/RecoveryManager.java

```diff
@@ -0,0 +1,367 @@
+     */
+    private Map<String, Xid> retrievePreparedXids(XAResource xaResource) throws XAException {
+        Map<String, Xid> retrievedXids = new HashMap<>();
+        ArrayList<Xid> recoverdXidsFromScan = new ArrayList<>();
+        
+        Xid[] xidsFromScan = null;
+        xidsFromScan = xaResource.recover(XAResource.TMSTARTRSCAN);
+        while (xidsFromScan != null && xidsFromScan.length > 0) {
+            recoverdXidsFromScan.addAll(List.of(xidsFromScan));
+            xidsFromScan = (xaResource.recover(XAResource.TMNOFLAGS));
+        }
+        xidsFromScan = xaResource.recover(XAResource.TMENDRSCAN);
+        if (xidsFromScan != null && xidsFromScan.length > 0) {
+            recoverdXidsFromScan.addAll(List.of(xidsFromScan));
+        }
+
+        if (!recoverdXidsFromScan.isEmpty()) {
+            for (Xid xid : recoverdXidsFromScan) {
+                if (xid == null) {
+                    continue;
+                }
+                if (xid.getFormatId() != (XIDGenerator.getDefaultFormat())){
+                    continue;
+                }
+                String globalTransactionIdStr = new String(xid.getGlobalTransactionId());
+                String branchQualifierStr = new String(xid.getBranchQualifier());
+                String combinedIdStr = globalTransactionIdStr + ":" + branchQualifierStr;
+                if (retrievedXids.containsKey(combinedIdStr)) {
+                    continue;
+                }
+                retrievedXids.put(combinedIdStr, xid);
+            }
+        }
+        return retrievedXids;
+    }
+
+    /**
+     * Handle commit of a transaction in transaction recovery
+     *
+     * @param xaResource the resource that the transaction is associated with
+     * @param xid the xid of the transaction
+     * @return true if the commit is successful and the log can be forgotten, false otherwise
+     */
+    private boolean replayCommit(XAResource xaResource, Xid xid) {
+        boolean ret = false;
+        try {
+            xaResource.commit(xid, false);
+            ret = true;
+        } catch (XAException e) {
+            switch (e.errorCode){
+                // case: transaction already heuristically terminated by resource
+                case XAException.XA_HEURCOM,
+                        XAException.XA_HEURHAZ,
+                        XAException.XA_HEURMIX,
+                        XAException.XA_HEURRB:
+                    ret = handleHeuristicTermination(xid, xaResource, e, true);
+                    break;
+                // case : transaction terminated in resource by a concurrent commit; xid no longer know by resource
+                case XAException.XAER_NOTA, XAException.XAER_INVAL:
+                    ret = true;
+                    break;
+                default:
+                    log.error("transient error while replaying commit for transaction: " + xid + " " + e.getMessage());
+            }
+        }
+        return ret;
```

Comment:
We can return in each switch cases separatatly. wdyt?

Reply:
fixed in [4fd4179](https://github.com/ballerina-platform/ballerina-lang/pull/42080/commits/4fd4179e55c2758cacf679862c6ca4a2988b36d8)